File: src/main/java/com/ultreon/bubbles/entity/AmmoEntity.java
package com.ultreon.bubbles.entity;

import com.ultreon.commons.annotation.FieldsAreNonnullByDefault;
import com.ultreon.commons.annotation.MethodsReturnNonnullByDefault;
import com.ultreon.bubbles.common.gametype.AbstractGameType;
import com.ultreon.bubbles.entity.ammo.AmmoType;
import com.ultreon.bubbles.environment.Environment;
import com.ultreon.bubbles.event.EntityCollisionEvent;
import com.ultreon.bubbles.init.Entities;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import javax.annotation.ParametersAreNonnullByDefault;
import java.awt.*;
import com.ultreon.hydro.render.Renderer;

@MethodsReturnNonnullByDefault
@FieldsAreNonnullByDefault
@ParametersAreNonnullByDefault
public class AmmoEntity extends Entity {
    @Nullable
    private AmmoType ammoType;
    private double rotation;

    public AmmoEntity(AbstractGameType gameType) {
        this(null, gameType);
    }

    public AmmoEntity(@Nullable AmmoType type, AbstractGameType gameType) {
        super(Entities.AMMO.get(), gameType);

        this.ammoType = type;
        if (type != null) {
            this.bases.setAll(type.getDefaultAttributes());
            this.attributes.setAll(type.getDefaultAttributes());
        }
    }

    @Override
    public void onCollision(EntityCollisionEvent evtObj) {

    }

    @Override
    public Shape getShape() {
        return ammoType != null ? ammoType.getShape(this) : new Rectangle(0, 0, 0, 0);
    }

    @Override
    public void renderEntity(Renderer gg) {
        if (this.ammoType != null) {
            this.ammoType.render(gg, this);
        }
    }

    @Override
    public void delete() {

    }

    @Override
    protected void bindEvents() {

    }

    @Override
    protected void unbindEvents() {

    }

    @Override
    protected boolean areEventsBound() {
        return false;
    }

    @Override
    public void tick(Environment environment) {

    }

    @SuppressWarnings("unused")
    @Nullable
    public AmmoType getAmmoType() {
        return ammoType;
    }

    @SuppressWarnings("unused")
    public void setAmmoType(@NotNull AmmoType ammoType) {
        this.ammoType = ammoType;
        this.bases.setAll(ammoType.getDefaultAttributes());
        this.attributes.setAll(ammoType.getDefaultAttributes());
    }

    public double getRotation() {
        return rotation;
    }

    public void setRotation(double rotation) {
        this.rotation = rotation;
    }
}

File: src/main/kotlin/qtech/bubbles/entity/AmmoEntity.kt
package qtech.bubbles.entity

import qtech.bubbles.annotation.FieldsAreNonnullByDefault
import qtech.bubbles.annotation.MethodsReturnNonnullByDefault
import qtech.bubbles.common.ammo.AmmoType
import qtech.bubbles.common.gametype.AbstractGameMode
import qtech.bubbles.environment.Environment
import qtech.bubbles.event.CollisionEvent
import qtech.hydro.GraphicsProcessor
import qtech.bubbles.init.Entities
import java.awt.Rectangle
import java.awt.Shape
import javax.annotation.ParametersAreNonnullByDefault

@MethodsReturnNonnullByDefault
@FieldsAreNonnullByDefault
@ParametersAreNonnullByDefault
class AmmoEntity(ammoType: AmmoType?, gameMode: AbstractGameMode) : DamageableEntity(Entities.AMMO.get(), gameMode) {
    var ammoType = ammoType
        set(value) {
            field = value
            bases.setAll(ammoType!!.defaultAttributes)
            attributeMap.setAll(ammoType!!.defaultAttributes)
        }
    public override var rotation = 0.0

    constructor(gameMode: AbstractGameMode) : this(null, gameMode)

    override fun onCollision(event: CollisionEvent) {}
    override val shape: Shape
        get() = if (ammoType != null) ammoType!!.getShape(this) else Rectangle(0, 0, 0, 0)

    override fun renderEntity(gg: GraphicsProcessor?) {
        if (ammoType != null) {
            ammoType!!.render(gg!!, this)
        }
    }

    override fun delete() {}
    override fun bindEvents() {}
    override fun unbindEvents() {}
    override fun areEventsBound(): Boolean {
        return false
    }

    override fun tick(environment: Environment?) {}
    init {
        if (ammoType != null) {
            bases.setAll(ammoType.defaultAttributes)
            attributeMap.setAll(ammoType.defaultAttributes)
        }
    }
}